tests/ui/lint/non_local_definitions.rs

```diff
@@ -0,0 +1,337 @@
+    }
+
+    impl Tait for super::Test {}
+}
+
+trait Uto {}
+const Z: () = {
+    trait Uto1 {}
+
+    impl Uto1 for Test {} // the trait is local, don't lint
+
+    impl Uto for &Test {}
+    //~^ WARN non local definition
+};
+
+trait Ano {}
+const _: () = {
+    impl Ano for &Test {} // ignored since the parent is a anon-const
+};
+
+type A = [u32; {
+    impl Uto for *mut Test {}
+    //~^ WARN non local definition
+
+    1
+}];
+
+enum Enum {
+    Discr = {
+        impl Uto for Test {}
+        //~^ WARN non local definition
+
+        1
+    }
+}
+
+trait Uto2 {}
+static A: u32 = {
+    impl Uto2 for Test {}
+    //~^ WARN non local definition
+
+    1
+};
+
+trait Uto3 {}
+const B: u32 = {
+    impl Uto3 for Test {}
+    //~^ WARN non local definition
+
+    #[macro_export]
+    macro_rules! m0 { () => { } };
+    //~^ WARN non local definition
+
+    trait Uto4 {}
+    impl Uto4 for Test {}
+
+    1
+};
+
+trait Uto5 {}
+fn main() {
+    #[macro_export]
+    macro_rules! m { () => { } };
+    //~^ WARN non local definition
+
+    impl Test {
+    //~^ WARN non local definition
+        fn foo() {}
+    }
+
+    let _array = [0i32; {
+        impl Test {
+        //~^ WARN non local definition
+            fn bar() {}
+        }
+
+        1
+    }];
+
+    const {
+        impl Test {
+        //~^ WARN non local definition
+            fn hoo() {}
+        }
+
+        1
+    };
+
+    impl Display for Test {
+    //~^ WARN non local definition
+        fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            todo!()
+        }
+    }
+
+    impl dyn Uto5 {}
+    //~^ WARN non local definition
+
+    impl<T: Uto5> Uto5 for Vec<T> { }
+    //~^ WARN non local definition
+
+    impl Uto5 for &dyn Uto5 {}
+    //~^ WARN non local definition
+
+    impl Uto5 for *mut Test {}
+    //~^ WARN non local definition
+
+    impl Uto5 for *mut [Test] {}
+    //~^ WARN non local definition
+
+    impl Uto5 for [Test; 8] {}
+    //~^ WARN non local definition
+
+    impl Uto5 for (Test,) {}
+    //~^ WARN non local definition
+
+    impl Uto5 for fn(Test) -> () {}
+    //~^ WARN non local definition
+
+    impl Uto5 for fn() -> Test {}
+    //~^ WARN non local definition
+
+    let _a = || {
+        impl Uto5 for Test {}
+        //~^ WARN non local definition
+
+        1
+    };
+
+    type A = [u32; {
+        impl Uto5 for &Test {}
+        //~^ WARN non local definition
+
+        1
+    }];
+
+    fn a(_: [u32; {
+        impl Uto5 for &(Test,) {}
+        //~^ WARN non local definition
+
+        1
+    }]) {}
+
+    fn b() -> [u32; {
+        impl Uto5 for &(Test,Test) {}
+        //~^ WARN non local definition
+
+        1
+    }] { todo!() }
+
+    struct InsideMain;
+
+    impl Uto5 for *mut InsideMain {}
+    impl Uto5 for *mut [InsideMain] {}
+    impl Uto5 for [InsideMain; 8] {}
+    impl Uto5 for (InsideMain,) {}
+    impl Uto5 for fn(InsideMain) -> () {}
+    impl Uto5 for fn() -> InsideMain {}
```

Comment:
All of those should warn too, they are no different to `UwU<OwO>` example down below: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=425c7e44f13465c39d440ca975765e20

Reply:
In other words I think `self_ty_has_local_parent` should just be `matches!(impl_.self_ty.kind, TyKind::Path(QPath::Resolved(_, ty_path)) if path_has_local_parent(ty_path))` or something like it. It's never safe to look into generics for this lint.

Reply:
Done. This simplifies many thing, thanks.